Fix cmus remote disconnections

kingo55 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When cmus is restarted after being initialized in Home Assistant, the integration breaks. This reconnects in case it can't reach cmus with the first connection.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes this issue described in Discourse: https://community.home-assistant.io/t/remote-cmus-integration-stops-working-when-cmus-is-stopped-restarted/223868

1. If we start HASS before cmus: cmus doesn't connect
2. If we start cmus, then HASS: cmus works
3. If cums is working in HASS and we restart cmus: cmus integration reports broken pipe

```
2020-09-19 11:57:07 WARNING (SyncWorker_3) [homeassistant.components.cmus.media_player] [Errno 32] Broken pipe
```


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @kingo55, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#40284).

Once you do that we will be able to review and accept this pull request.

Thanks!

mdonoughe commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -86,11 +86,38 @@ def __init__(self, server, password, port, name):
+        self._port = port
         self.status = {}
 
+    def reconnect(self):
```
This reconnect method invokes `remote.PyCmus()` within a try block, allowing it to fail. `__init__` executes the same code without a try block. Probably they should use the same code so the integration doesn't fail to start if the first connection attempt fails, and for less code duplication.

kingo55 replied:
Good call - just tidied this up and it feels a bit tighter now.

MartinHjelmare replied:
I'd move this method to a new class that when instantiated and connected has the cmus attribute. Then pass an instance of this new class to the entity when creating it.

mdonoughe commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -86,11 +86,38 @@ def __init__(self, server, password, port, name):
+                self.cmus = remote.PyCmus()
+        except exceptions.InvalidPassword:
+            _LOGGER.error("The provided password was rejected by cmus")
+            return False
```
When an exception is thrown, this method returns a boolean which is not used. When no exception is thrown, this method doesn't return anything.

kingo55 replied:
Thanks.

mdonoughe commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -86,11 +86,38 @@ def __init__(self, server, password, port, name):
+        except BrokenPipeError:
+            self.reconnect()
+            status = self.cmus.get_status_dict()
+        except exceptions.ConfigurationError:
```
When this exception occurs, reconnect is called, but status is not set like in the previous block. I think this will cause `NameError` for `if not status` because `status` is never set.

kingo55 replied:
Since we attempt a reconnection here, the status should be accessible if it successfully reconnects.

Though since reconnections are usually from the cmus server being off, I think this might be sensible not to try updating the status here.

kingo55 commented:
Awesome... thankyou for reviewing my tweaks @mdonoughe !

They all make sense - I need to go ahead and make those updates / test it's working OK. Will let you know if I have any questions.

kingo55 commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -80,18 +80,41 @@ def __init__(self, server, password, port, name):
+        except exceptions.ConfigurationError:
+            self.connect()
+            _LOGGER.warning("A configuration error occurred")
+        if "status" not in locals():
```
Safely check the status is set (in case it's not defined here). `if not status` was throwing exceptions here.

kingo55 commented:
Not sure what else I can do here... need more reviews to fix this problem.

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -80,18 +80,41 @@ def __init__(self, server, password, port, name):
+        self._password = password
+        self._port = port
         self.status = {}
+        self.connect()
```
We don't want side effects in init methods. Please create and set up the remote device first and then pass that instance to the entity when creating that.

Possibly wrap the device api inside a new class that has a (re)connect method to refresh the connection. It's good to separate the entity api from the device api and not add device specific methods to the entity.

kingo55 replied:
Thanks @MartinHjelmare ! It's getting late here - I'll try implementing this on the weekend.

kingo55 replied:
@MartinHjelmare - I managed to remove the `connect()` method from `__init()__`.

Rather than wrapping it up inside a new class, I just passed the class from the `remote.PyCmus` package. It should mean a bit less code.

Also, I created a new `reconnect()` function to handle the disconnects within the main device's class.

Was this what you had in mind?

kingo55 committed:
fix: cmus reconnect remote on disconnection

kingo55 commented:
Sorry guys, I nuked my history while rebasing.

Since it's quite old, I wanted to make sure it worked on the latest version of the dev branch and the new unit tests etc.

kingo55 committed:
Move cmus connection to separate class

kingo55 commented:
@MartinHjelmare - sorry, my Python isn't very good.

I think I get what you mean now, though. I broke it out into a separate class and I think it's allowed me to simplify the connection. Also, there's no reconnection function in the CmusDevice anymore.

Is this what you wanted?

kingo55 committed:
Address flake tests

kingo55 commented:
Looks like my tox isn't running the same pylint tests locally as github is (this computer is missing Python 3.8 apparently :facepalm:). 

The latest commit addresses:

* Redefinition of `self.remote`
* Defining `self.cmus` outside the `__init__` context

I know you asked not to add this to `__init__` but I was't sure if that was specifically for the CmusDevice class. Please let me know if this is wrong or if there might be another way to handle it.

kingo55 committed:
Address additional flake tests

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -65,33 +65,67 @@ def setup_platform(hass, config, add_entities, discover_info=None):
+        cmus_remote = CmusRemote(server=host, port=port, password=password)
     except exceptions.InvalidPassword:
         _LOGGER.error("The provided password was rejected by cmus")
         return False
```
Nothing is checking this return value.
```suggestion
        return
```

kingo55 replied:
Cheers - going to include this in my next commit.

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -65,33 +65,67 @@ def setup_platform(hass, config, add_entities, discover_info=None):
+        self._server = server
+        self._port = port
+        self._password = password
+        self.cmus = self.connect()
```
Please don't do side effects in init methods. Call connect explicitly after creating the instance.

kingo55 replied:
OK, perhaps if I define it as None and reassign during connect, that should do the trick?

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -65,33 +65,67 @@ def setup_platform(hass, config, add_entities, discover_info=None):
+        except exceptions.InvalidPassword:
+            _LOGGER.error("The provided password was rejected by cmus")
+
+        return cmus
```
This will error if the exception is triggered. Then `cmus` won't be defined.

kingo55 replied:
Thanks - good catch.

kingo55 committed:
Move cmus connection outside of __init__

kingo55 commented:
@MartinHjelmare - I refactored this so we don't need to call `connect()` during init. This also changes the behaviour of `connect()` so we no longer need a `reconnect()` function.

Finally, I added a check to see that cmus was connected before passing it into `add_entities()`, too. Seems like the right thing to do if it's not connected at startup, right?

Thanks for your review and bearing with this PR here.

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -65,33 +65,66 @@ def setup_platform(hass, config, add_entities, discover_info=None):
+        if cmus_remote.cmus is None:
+            return
+
     except exceptions.InvalidPassword:
```
This will never happen now since we catch the error inside the `CmusRemote` class.

kingo55 replied:
Good point! I just removed the redundant try-catch here. Much cleaner...

kingo55 committed:
Remove redundant exception handler

MartinHjelmare commented on homeassistant/components/cmus/media_player.py:
```diff
@@ -64,34 +64,62 @@ def setup_platform(hass, config, add_entities, discover_info=None):
+        except BrokenPipeError:
+            self._remote.connect()
+        except exceptions.ConfigurationError:
+            self._remote.connect()
+            _LOGGER.warning("A configuration error occurred")
+        if "status" not in locals():
             _LOGGER.warning("Received no status from cmus")
         else:
             self.status = status
```
```suggestion
            _LOGGER.warning("A configuration error occurred")
            self._remote.connect()
        else:
            self.status = status
            return

        _LOGGER.warning("Received no status from cmus")
```

MartinHjelmare committed:
Clean up

MartinHjelmare reviewed (approved):
Thanks!
